janusgraph-core/src/main/java/org/janusgraph/graphdb/query/index/ApproximateIndexSelectionStrategy.java

```diff
@@ -0,0 +1,102 @@
+import org.janusgraph.graphdb.internal.OrderList;
+import org.janusgraph.graphdb.query.condition.Condition;
+import org.janusgraph.graphdb.query.condition.MultiCondition;
+import org.janusgraph.graphdb.query.graph.JointIndexQuery;
+import org.janusgraph.graphdb.types.IndexType;
+import org.janusgraph.graphdb.types.MixedIndexType;
+
+/**
+ * @author Florian Grieskamp (Florian.Grieskamp@gdata.de)
+ */
+public class ApproximateIndexSelectionStrategy
+    extends AbstractIndexSelectionStrategy {
+
+    /**
+     * Iterate over all potential indexes and compute a score based on how many clauses
+     * this index covers. The index with the highest score (as long as it covers at least one
+     * additional clause) is picked and added to the joint query for as long as such exist.
+     */
+    @Override
+    public SelectedIndexQuery selectIndices(final Set<IndexType> rawCandidates,
+                                            final MultiCondition<JanusGraphElement> conditions,
+                                            final Set<Condition> coveredClauses, OrderList orders,
+                                            IndexSerializer serializer) {
+        final JointIndexQuery jointQuery = new JointIndexQuery();
+        boolean isSorted = orders.isEmpty();
+        while (true) {
+            IndexCandidate bestCandidate = null;
+            boolean candidateSupportsSort = false;
+
+            for (final IndexType index : rawCandidates) {
+                final IndexCandidate indexCandidate =
+                    createIndexCandidate(index, conditions, serializer);
+                if (indexCandidate == null) {
+                    continue;
+                }
+
+                boolean supportsSort =
+                    orders.isEmpty() ||
+                    coveredClauses.isEmpty() && index.isMixedIndex() &&
+                        IndexSelectionUtil.indexCoversOrder((MixedIndexType) index, orders);
+                indexCandidate.setScore(calculateIndexCandidateScore(indexCandidate, coveredClauses, supportsSort));
+
+                if (!coveredClauses.containsAll(indexCandidate.getSubCover()) &&
+                    (bestCandidate == null ||
+                     indexCandidate.getScore() > bestCandidate.getScore())) {
+                    bestCandidate = indexCandidate;
+                    candidateSupportsSort = supportsSort;
+                }
+            }
+
+            if (bestCandidate != null) {
+                if (coveredClauses.isEmpty()) {
+                    isSorted = candidateSupportsSort;
+                }
+                coveredClauses.addAll(bestCandidate.getSubCover());
+                addToJointQuery(bestCandidate, jointQuery, serializer, orders);
+            } else {
+                break;
+            }
+        }
+        return new SelectedIndexQuery(jointQuery, isSorted);
+    }
+
+    private double calculateIndexCandidateScore(final IndexCandidate indexCandidate,
```

Comment:
It would be better to pass only necessary parameters to the method. Even so, we will need to pass 2 parameters (subCover and index), it is better to pass them instead of indexCandidate because `indexCandidate` has more details (which are not needed by this method, i.e. subCondition)

Reply:
I think it's convenient to pass the entire `IndexCandidate` object to the score calculation. Especially as we trimmed down the class `IndexCandidate` to contain only relevant information and no logic, in my opinion every attribute of of an `IndexCandidate` qualifies as a valid foundation to be considered for the calculation of a score.

Reply:
Good point. Agree